File: w5/2/proxy.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <winsock.h>

#define MAXLINE 1024 

int main(){
    /*
        Data which will be send/resv to DNS Serv and resv from time Server
    */
    char  str[MAXLINE];

    /*
        Register Winsock API
        Amd Make Sure Whether WinSock API Can Use
    */
    WSADATA wsadata;
    if (WSAStartup(0x101,(LPWSADATA) &wsadata) != 0) {
        fprintf(stderr,"can't use WinSock DLL\n");
        exit(1);
    }
    /*
        Declare Socket
        And Make Sure TCP socket Could be Use or not
    */
    SOCKET ServSock;
    if ( (ServSock=socket(AF_INET, SOCK_STREAM, 0)) == SOCKET_ERROR) {
        fprintf(stderr," can't open TCP socket\n");
        exit(1);
    }
    /*
        Server Information
        ProxyServer
            addr: 127.0.0.1
            port: 1234
        DateTimeServer
            addr: 127.0.0.1
            port: 13
    */
    struct sockaddr_in serv;
    serv.sin_family = AF_INET;
    serv.sin_port = htons(1234);
    serv.sin_addr.s_addr = inet_addr("127.0.0.1");

    struct sockaddr_in DateTimeServer;
    DateTimeServer.sin_family = AF_INET;
    DateTimeServer.sin_port = htons(13);
    DateTimeServer.sin_addr.s_addr = inet_addr("127.0.0.1");
    /*
        Bind
    */
    bind(ServSock, (struct sockaddr *) &serv, sizeof(serv));


    struct sockaddr_in client;
    int clnt_len=sizeof(client);
    listen(ServSock,5);
    while(1){
        
        printf("Proxy Server Wait\n");

        /*
            Get Connect from clinet
        */
       SOCKET ClientSock;
       ClientSock =  accept(ServSock, (struct sockaddr *)  &client,&clnt_len );
       while(1){
            char meun[]= "Please enter the Service u want to use\n1)IP Query\n2)Datetime Service\n";
            if (send(ClientSock,meun, strlen(meun)+1,0) == SOCKET_ERROR){
                closesocket(ClientSock);
                break;
            } 
            int size = recv(ClientSock,str, 1024, 0);
            if (size < 0){
                closesocket(ClientSock);
                break;
            }
            str[size] = '\0';
            if (strcmp(str,"1") == 0){
                if (send(ClientSock,"Please enter a IP address\n", strlen("Please enter a IP address\n")+1,0)== SOCKET_ERROR){
                    closesocket(ClientSock);
                    break;
                }

                if (recv(ClientSock,str, 1024, 0)< 0){
                    closesocket(ClientSock);
                    break;
                }
                sscanf(str, "ask://%s", &str);
                LPHOSTENT hp;
                struct in_addr sAddr;
                sAddr.s_addr=inet_addr(str);
                hp= gethostbyaddr((LPSTR) &sAddr, sizeof(sAddr),AF_INET);
                if (hp == NULL){
                    if(send(ClientSock,"NOT FOUND\n",strlen("NOT FOUND\n")+1,0) == SOCKET_ERROR){
                        closesocket(ClientSock);
                        break;
                    }
                }
                else{
                    if(send(ClientSock,hp->h_name,strlen(hp->h_name)+1,0) == SOCKET_ERROR){
                        closesocket(ClientSock);
                        break;
                    }
                }
            }
            else if (strcmp(str,"2") == 0){
                SOCKET ProxySock;
                if ( (ProxySock=socket(AF_INET, SOCK_STREAM, 0)) == SOCKET_ERROR) {
                    fprintf(stderr," can't open TCP socket\n");
                    exit(1);
                }
                connect(ProxySock, (LPSOCKADDR) &DateTimeServer, sizeof(DateTimeServer));
                int size = recv(ProxySock,str, 1024, 0);
                if (size < 0){
                    closesocket(ClientSock);
                    break;
                }
                if(send(ClientSock,str,strlen(str)+1,0) == SOCKET_ERROR){
                    closesocket(ClientSock);
                    break;
                }
                closesocket(ProxySock);
            }
            else{
                if(send(ClientSock,"Invalid data!\n", strlen("Invalid data!\n")+1,0) == SOCKET_ERROR){
                    closesocket(ClientSock);
                    break;
                }
            }
            
       }

       
       closesocket(ClientSock);
    }
}

File: w5/2/proxy.py
import socket

"""
Set up Proxy Server information
    Proxy Server
        IP:   127.0.0.1
        Port: 1234
"""
ProxyIP   = '127.0.0.1'
ProxyPort = 1234

"""
Set up Server Information
    DNS Server
        we implement by `gethostbyaddr` funtion
    DateTime Server
        IP:   127.0.0.1
        Port: 13
"""
DateTimeServerIP   = "127.0.0.1"
DateTimeServerPort = 13

"""
Socket setup 
    Server Socket
        For accept client socket
        varible: servsd
    Client Socket
        For Query Date time
        Varible: clintsd
"""
servsd  = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
clintsd = socket.socket(socket.AF_INET, socket.SOCK_STREAM)


"""
bind Server
"""
servsd.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
servsd.bind((ProxyIP, ProxyPort))
servsd.listen(5)

"""
Excute Proxy Server
    1. accept client conntion
    2. resv client want to use which service
    3. send data to server
    4. send data to client
"""
while True:
    print('wait for connection...')
    conn, addr = servsd.accept()
    print('connected by ' + str(addr))
    while True:
        resvData = conn.recv(1024).decode()
        if resvData == "1":
            print("Client Select DNS Query")
            #DNS query
            resvData = list(map(int,(conn.recv(1024).decode()[6:]).split('.')))
            #Enforce data integrity
            if len(resvData) != 4 or not (0<=resvData[0]<=255) or not(0<=resvData[1]<=255) or not(0<=resvData[2]<=255) or not(0<=resvData[3]<=255):
                conn.send("Invalid Data".encode())
                continue
            ip = str(resvData[0]) + "." + str(resvData[1]) + "." + str(resvData[2]) + "." + str(resvData[3])
            hostName = socket.gethostbyaddr(ip)
            if hostName == None:
                print("proxyget: NOTFOUND")
                conn.send("NOTFOUND".encode())
                print("proxysend: NOTFOUND")
            else:
                print(f"proxyget:{str(hostName[0])}")
                conn.send(str(hostName[0]).encode())
                print(f"proxysend:{str(hostName[0])}")
        elif resvData == "2":
            #DateTime Service
            clintsd = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            clintsd.connect((DateTimeServerIP,DateTimeServerPort))
            #Remember that, server who sending data use big5 encoding, but decode() use uft8 initiality  
            data = clintsd.recv(1024).decode("big5")
            print(f"proxyget:{data}")
            clintsd.close()
            conn.send(data.encode())
            print(f"proxysend:{data}")
        else:
            #Ivalid Data
            conn.send("Invalid Data".encode())
    print("=====================")
servsd.close()